File: cpu4/c/reorder_buffer.c
struct rob_entry {
  int waiting;
  int opcode;
  int tag;
  int value;
  int flags;
};

struct reorder_buffer {
  int head;
  int tail;
  struct rob_entry entries[32];
};

/* Initialize the reorder buffer. */

void
rob_init(struct reorder_buffer *rob) {
  int i;
#define ENTRY (rob->entries[i])
  for (i = 0; i < 32; i++) {
    ENTRY.waiting = 0;
  }
  head = 0;
  tail = 0;
#undef ENTRY
}

/* Push a new instruction to the ROB. */

int
rob_push(struct reorder_buffer *rob, int opcode, int tag) {
  if (rob->head + 1 == rob->tail)
    return 0;
#define HEAD (rob->entries[rob->head])
  HEAD.waiting = 1;
  HEAD.opcode = opcode;
  HEAD.tag = tag;
#undef HEAD
  rob->head = (rob->head + 1) % 32;
  return 1;
}

/* Get the next instruction to retire from the commit queue. */

int
rob_next(struct reorder_buffer *rob, int *opcode) {
  if (rob->tail == rob->head)
    return 0;
#define TAIL (rob->entries[rob->tail])
  if (TAIL.waiting)
    return 0;
  *opcode = TAIL.opcode;
#undef TAIL
  rob->tail = (rob->tail + 1) % 32;
  return 1;
}

/* Lookup an operand by tag. */

int
rob_lookup(struct reorder_buffer *rob, struct operand *operand) {
  
}

File: cpu4/python/reorder_buffer.py
import copy
from packed_instruction import *

# Tags are basically modular arithmetic, so we need a special operation to
# properly compute the offset in the reorder buffer.

def tag_diff(a, b):
  return a - b

# Use this placeholder class for storing reorder buffer entries.

class ROB_Entry:
  def __init__(self):
    self.instr = PackedInstruction()
    self.waiting = False

class ReorderBuffer:
  def __init__(self):
    self.entries = [ROB_Entry() for i in range(4)]
    self.head, self.tail = 0, 0

  def push(self, instr):
    entry = self.entries[self.head]
    entry.instr = copy.deepcopy(instr)
    entry.waiting = True

  # Lookup an entry in the reorder buffer by tag, if it's still pending then
  # return None

  def lookup(self, tag, i):
    base_entry = self.entries[0]
    base_tag = base_entry.instr.op_tags[i]
    diff = tag_diff(tag, base_tag)
    if (diff > self.head):
      return None
    entry = self.entries[diff]
    if (entry.waiting):
      return None
    return entry.instr.result

  # Writeback to an entry in the reorder buffer and set the waiting bit to
  # False

  def writeback(self, tag, result):
    base_entry = self.entries[0]
    base_tag = base_entry.instr.wb_tag
    diff = tag_diff(tag, base_tag)
    if (diff > self.head):
      return None
    entry = self.entries[diff]
    entry.waiting = False
    entry.instr.result = result
